### PR TITLE
Add Speech Commands to NeMo

### PR DESCRIPTION
**Pull request by titu1994**
# Changelog

## Additions
- Add `kernel_width`, `se`, and `se_reduction_ratio` parameters to `JasperBlock` to support conv kernel width alteration and squeeze-and-excitation sub-modules.
- Add `JasperDecoderForClassification`
- Add `CropOrPadSpectrogramAugmentation` from OpenSeq2Seq
- Add `AudioToSpeechLabelDataLayer` and supporting collection and Dataset
- Add `CrossEntropyLossNM`
- Add `WhiteNoisePerturbation`
- Add `WarmupHoldPolicy`, `PolynomialDecayAnnealing` and `PolynomialHoldDecayAnnealing` learning rate policies
- Add logging callbacks and helpers
- Add "speech_command" section to docs
- Add `scripts/process_speech_commands_data.py` for dataset download and manifest building
- Add `examples/asr/quartznet_speech_commands.py` training script as well as config files for Google Speech Commands dataset v1 and v2.
 - Add tutorial in docs for training, inference and evaluation of results (listen to incorrect cases)

## Changes
- Refactor `manifest.iter_item` to accept an optional Callable which will be used to parse each line of the manifest. When no default Callable is provided, internally will switch to use the private ASR parser within that module.
- Check `isinstance(optimizer, torch.optim.**O**ptimizer)` when passing instantiated optimizer instead of string to Factory otherwise python complains that you cannot test against a module `torch.optim.**o**ptimizer`

Signed-off-by: smajumdar <titu1994@gmail.com>

### CODE REVIEW

File: nemo/collections/asr/losses.py [hunk 1]
```diff
@@ -58,3 +59,9 @@ def _loss(self, log_probs, targets, input_length, target_length):
 
     def _loss_function(self, **kwargs):
         return self._loss(*(kwargs.values()))
+
+
+class CrossEntropyLossNM(losses.CrossEntropyLoss):
```
Comment (blisc) [inline nemo/collections/asr/losses.py]: We already have crossentropyloss defined elsewhere, please use that

File: nemo/collections/asr/data_layer.py [hunk 2]
```diff
@@ -425,3 +428,165 @@ def dataset(self):
...
+            'max_duration': max_duration,
+            'min_duration': min_duration,
+            'trim': trim_silence,
+            'load_audio': load_audio,
+        }
+        self._dataset = AudioLabelDataset(**dataset_params)
+
+        # Set up data loader
+        if self._placement == DeviceType.AllGpu:
+            nemo.logging.info("Parallelizing Datalayer.")
```
Comment (blisc) [inline nemo/collections/asr/data_layer.py]: Please switch all instances of nemo.logging to just logging and adding either `logging = nemo.logging` or `from nemo import logging`. Can you also change all 12 instances? But there is no need to change what was already there. I will change it in my PR.

File: nemo/collections/asr/losses.py (hunk 1 above)
Comment (titu1994) [reply nemo/collections/asr/losses.py]: Good point, added alias in asr __init__.py to get the backend function. Added "NM" to it so that it is clear that this is a NM and not pytorch loss.

File: nemo/collections/asr/data_layer.py (hunk 2 above)
Comment (titu1994) [reply nemo/collections/asr/data_layer.py]: Will change nemo.logging to logging. However, 12 instances of nemo.logging? Not taking into account the earlier data layers, I can see only 2 logging commands in AudioToSpeechDataLayer. Does the 12 refer to changes on all of the files?

Another thing is do I keep both logging and nemo.logging for files that were already there?

Comment (blisc) [reply nemo/collections/asr/data_layer.py]: You don't have to change any logging statements that are already present. But for the new ones that you added, can you change them?
There should be 12 across all changed files.

File: docs/sources/source/speech_command/datasets.rst [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+Google Speech Commands Dataset
+-----------------------------------
+
+The ability to recognize spoken commands with high accuracy can be useful in a variety of contexts.
+To this end, Google released the Speech Commands dataset (see :cite:`speech-recognition-dataset-warden2018speech`,
+which contains short audio clips of a fixed number of command words such as “stop”, “go”, “up”, “down”, etc spoken by a large number of speakers.
+To promote the use of the set, Google also hosted a Kaggle competition, in which the winning team attained a multiclass accuracy of 91%.
+
+We experimented with applying NeMo’sYes ill jus ASR classification models on mel spectrograms of the audio clips and found that they worked surprisingly well.
```
Comment (vsl9) [inline docs/sources/source/speech_command/datasets.rst]: Looks like a typo

File: docs/sources/source/speech_command/quartznet.rst [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+`QuartzNet paper <https://arxiv.org/abs/1910.10261>`_.
+
+These QuartzNet models were trained for 200 epochs using mixed precision on 2 GPUs with a batch size of 128 over 200 epochs.
+On 2 Quadro GV100 GPUs, training time is approximately 1 hour.
+
+=============================== ===================== ============
+Network                         Dataset               Results
+=============================== ===================== ============
+QuartzNet3x1 (0.077M params)    Speech Commands V1    97.46% Test
```
Comment (vsl9) [inline docs/sources/source/speech_command/quartznet.rst]: Why not to use 77k instead of 0.077M?

File: docs/sources/source/speech_command/quartznet.rst [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+QuartzNet3x2 (0.093M params)    Speech Commands V2    97.35% Test
```
Comment (vsl9) [inline docs/sources/source/speech_command/quartznet.rst]: 93k vs 0.093M?

File: nemo/utils/lr_policies.py [hunk 6]
```diff
@@ -162,6 +221,42 @@ def _get_lr(self, initial_lr, step, epoch):
         return out_lr
 
 
+class PolynomialDecayAnnealing(WarmupPolicy):
```
Comment (vsl9) [inline nemo/utils/lr_policies.py]: Why not to add new policies to __all__?

File: nemo/collections/asr/jasper.py [hunk 7]
```diff
@@ -44,6 +44,21 @@ class JasperEncoder(TrainableNM):
...
                         # sub-blocks.
                         # Defaults to False
+                    'se' (bool)  # Whether to add Squeeze and Excitation
+                        # sub-blocks.
+                        # Defaults to False
+                    'se_reduction_ratio' (int)  # The reduction ratio of the Squeeze
+                        # sub-module.
+                        # Must be an integer > 1.
+                        # Defaults to 16
+                    'kernel_width' (float)  # Conv kernel size multiplier
```
Comment (vsl9) [inline nemo/collections/asr/jasper.py]: Maybe it is worth to rename it to something like 'kernel_size_factor', 'kernel_width_factor', 'width_factor'? In my experience, 'kernel_size' and 'kernel_width' are synonyms for 1D convolution. Here it is definitely a factor to scale the kernel size.

File: docs/sources/source/speech_command/datasets.rst (hunk 3 above)
Comment (titu1994) [reply docs/sources/source/speech_command/datasets.rst]: This and below issues fixed in latest commit

**Comment by blisc** [comment]
Can you add a line to the changelog?

File: nemo/utils/lr_policies.py (hunk 6 above)
Comment (titu1994) [reply nemo/utils/lr_policies.py]: Added in latest commit.

File: nemo/collections/asr/jasper.py (hunk 7 above)
Comment (titu1994) [reply nemo/collections/asr/jasper.py]: Changed to `kernel_size_factor` to remain in line with `kernel_size` naming convention.

File: nemo/collections/asr/helpers.py [hunk 8]
```diff
@@ -163,3 +200,86 @@ def post_process_predictions(predictions, labels):
...
+    batch_sizes = global_vars['batchsize']
+    total_num_samples = torch.tensor(batch_sizes).sum().float()
+
+    topk_accs = []
+    for k in top_k:
+        correct_counts = torch.tensor(global_vars[f'CorrectCount@{k}'])
+        topk_acc = correct_counts.sum() / total_num_samples
+        topk_accs.append(topk_acc)
+
+    if tag is None:
```
Comment (nithinraok) [inline nemo/collections/asr/helpers.py]: you can replace 
`tag='' if tag is None ` and use 

        logs = {f"Evaluation_Loss {tag}": eloss}
        logging.info(f"==========>>>>>>Evaluation Loss {tag}: {eloss}")
        for k, acc in zip(top_k, topk_accs):
            logging.info(f"==========>>>>>>Evaluation Accuracy Top@{k} {tag}: {acc * 100.:3.4f}")
            logs[f'Evaluation_Accuracy_Top@{k} {tag}'] = acc * 100.0

        return logs

without if else

File: examples/asr/quartznet_speech_commands.py [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+        num_workers=cpu_per_traindl,
+        augmentor=audio_augmentor,
+        **train_dl_params,
+    )
+
+    crop_pad_augmentation = nemo_asr.CropOrPadSpectrogramAugmentation(audio_length=128)
+
+    N = len(data_layer)
+    steps_per_epoch = math.ceil(N / (args.batch_size * args.iter_per_step * args.num_gpus))
+    print("Steps per epoch :", steps_per_epoch)
```
Comment (nithinraok) [inline examples/asr/quartznet_speech_commands.py]: Shouldn't print statement here replaced by nemo.logging ?

File: nemo/collections/asr/helpers.py (hunk 8 above)
Comment (titu1994) [reply nemo/collections/asr/helpers.py]: Fixed in latest commit

File: examples/asr/quartznet_speech_commands.py (hunk 9 above)
Comment (titu1994) [reply examples/asr/quartznet_speech_commands.py]: Fixed in latest commit.

**Comment by tu1258** [comment]
Hi, @titu1994 
Since Google Speech Command Dataset testing method is different from this. It choose 10 commands and "unknown word" and "silence" as label (look [section 7.1](https://arxiv.org/pdf/1804.03209.pdf)). And matchboxnet is excluded in keyword spotting discussion because of that (look [section 4.1 and 4.2](https://arxiv.org/pdf/2005.06720.pdf)). Would you mind also post this test benchmark and open source it?  Much thanks!

**Comment by titu1994** [comment]
Hello @tu1258,
Google speech commands can be tested in two ways, with the original 12 class challenge or with the all class challenge. We provide results on the all class problem and compare against other papers which also reported scores on the all class task.

We have also recently trained and evaluated MatchboxNet on the 12 class dataset. We must note that due to the non-deterministic way the tensorflow code you list generates the "silence" label, as well as shuffles the rest of the class labels, the dataset generated by pytorch and tensorflow codebases will not be the same. We will report the scores on this dataset in an update to the arxiv paper, and will notify you in this thread when it is updated.
